File: hard/0239_sliding_window_maximum/solution.cpp
#include <queue>
#include <vector>

class Solution {
public:
  std::vector<int> maxSlidingWindow(std::vector<int> &nums, int k) {
    std::deque<size_t> dq;
    std::vector<int> maxNums;

    for (size_t i{}; i < nums.size(); ++i) {
      while (!dq.empty() && nums[dq.back()] < nums[i])
        dq.pop_back();

      dq.push_back(i);

      while (!dq.empty() && i >= k && dq.front() <= i - k)
        dq.pop_front();

      if (i >= k - 1)
        maxNums.push_back(nums[dq.front()]);
    }

    return maxNums;
  }
};

File: hard/0239_sliding_window_maximum/solution.py
from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        queue = deque()
        max_nums = []

        for i, num in enumerate(nums):
            if queue and queue[0] <= i - k:
                queue.popleft()

            while queue and nums[queue[-1]] <= num:
                queue.pop()

            queue.append(i)
            max_nums.append(nums[queue[0]])

        return max_nums[k - 1 :]
